substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/sources/SourceManager.java

```diff
@@ -53,39 +51,35 @@ public class SourceManager {
      *         generated debug info or null if a source file cannot be found or cached.
      */
     public Path findAndCacheSource(ResolvedJavaType resolvedType, Class<?> clazz, DebugContext debugContext) {
-        /* short circuit if we have already seen this type */
-        Path path = verifiedPaths.get(resolvedType);
-        if (path != null) {
-            return (path != INVALID_PATH ? path : null);
+        // only instance classes and interfaces have a source file
+        if (!resolvedType.isInstanceClass() && !resolvedType.isInterface()) {
+            return null;
         }
-
+        // see if we can find a base file name for the class
         String fileName = computeBaseName(resolvedType);
-        /*
-         * null for the name means this class will not have a source so we skip on that
-         */
-        if (fileName != null) {
-            /*
-             * we can only provide sources for known classes and interfaces
-             */
-            if (resolvedType.isInstanceClass() || resolvedType.isInterface()) {
-                String packageName = computePackageName(resolvedType);
-                path = locateSource(fileName, packageName, clazz);
-                if (path == null) {
-                    // as a last ditch effort derive path from the Java class name
-                    if (debugContext.areScopesEnabled()) {
-                        debugContext.log(DebugContext.INFO_LEVEL, "Failed to find source file for class %s\n", resolvedType.toJavaName());
-                    }
-                    if (packageName.length() > 0) {
-                        path = Paths.get("", packageName.split("\\."));
-                        path = path.resolve(fileName);
-                    }
-                }
+        if (fileName == null) {
+            return null;
+        }
+        String packageName = computePackageName(resolvedType);
+        Path path = computePrototypeName(fileName, packageName);
+
+        /* lookup the source and race with other threads to cache it */
+        Source source = sourceIndex.computeIfAbsent(path, Source::new);
+        if (source.shouldCache()) {
+            // ask the source cache to locate the source file
+            String moduleName = null;
+            if (clazz != null) {
+                /* Paths require the module name as prefix */
+                moduleName = clazz.getModule().getName();
             }
+
+            cache.resolve(source, moduleName);
+
+            assert !source.isCaching();
         }
-        /* memoize the lookup */
-        verifiedPaths.put(resolvedType, (path != null ? path : INVALID_PATH));
 
-        return path;
+        /* cache attempt has been tried - return any associated path */
+        return source.getPath();
```

Comment:
What's the point of blocking before returning this value? Even if the source is still in the "CACHING" state does it really matter?